feat(doctor): separate doctor agents to specialize on a conflict type

honghanhh opened the pull request:
## Fixes

<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, open one before creating this pull request. -->

Fixes #95 #97

## Description

<!-- Concisely describe what the pull request does. -->
This PR focuses on split the Doctor Agents into 4 sub Doctor Agents, each analyzes clinical document pairs to determine how to introduce specific conflict types:
- pre/post care evolution
- temporality
- clinical history antecedents
- biological marker monitoring

Doctor agents use same proposition results from proposition agent, but each agent chooses different proposition pair which best suits its conflict type.

## Technical details

<!-- Add any other information or technical details about the implementation; or delete the section entirely. -->
Run the following command to quick generate top conflict output from 8 subsets.
```
python -m conflicts.run_top2
```
Flow:
 1. Setup: Init Hydra config and Pipeline (loads proposition agent, doctor agents, validators).                                                                                         
  2. Load data: Reads top2_per_conflict_type.parquet — pre-selected top-2 docs per conflict type.
  3. Per conflict type: Groups rows by conflict type, builds a DocumentPair from the first two rows.
  4. Proposition extraction: Runs proposition_agent on each document independently.
  5. Single doctor agent: Runs only the matching doctor agent (not all), returns attempts with retries/validation.
  6. Select best: Picks the valid attempt with the highest overall_score, or falls back to the last attempt.
  7. Save: Persists results to database and exports all results to JSON.

## Definition of Done

<!-- Check the boxes (replace the [ ] with [x]) corresponding to what you've done. -->
<!-- You must have followed the repository practices, but the other items can be skipped if your work doesn't require them. -->

- [x] I followed the repository practices (e.g. I used gitmojis).
- [ ] I have added/updated tests according to my changes.
- [ ] I have added/updated documentation according to my changes.

honghanhh committed:
feat: create 2 seperate doctor and edit agents for 2 new conflict types

honghanhh committed:
feat: update doctor and edit prompt

honghanhh committed:
revert: merge 2 edit agent into 1

honghanhh committed:
refactor: restructure  pipeline and doctor agent

honghanhh committed:
fix: add 2 new conflict types

honghanhh committed:
fix: remove deadcode, consolidate doctor agent, improve error handling

honghanhh committed:
chore: remove tested .json

honghanhh commented:
/gemini review

honghanhh committed:
refactor: simplify doctor agent and pipeline, rename proposition_pairs to proposition_conflicts
- Remove unnecessary helper methods and inline code
- Make proposition parameters required in DoctorAgent
- Rename proposition_pairs to proposition_conflicts for consistency
- Optimize statistics calculation to avoid recalculation
- Improve early exit logic in pipeline

honghanhh committed:
refactor: inline decompose_document_pair method, remove total_propositions field, use len(propositions) instead

honghanhh committed:
feat: add run_top2 script for targeted conflict type evaluation

honghanhh committed:
refactor: clean up run_top2

lenajoyeux-star commented on lib/conflicts/conflicts/core/pipeline.py:
```diff
@@ -145,89 +128,290 @@ def _save_attempts_to_database(
+            # Update best result from this conflict type's attempts
+            best_result = self._update_best_result(best_result, attempts)
+
+            # Early exit: Check success ngay sau khi process conflict type
```
🇻🇳 ❤️ 🇻🇳 ❤️ 🇻🇳 ❤️ 🇻🇳 ❤️ 🇻🇳 ❤️ 🇻🇳 ❤️ 🇻🇳 ❤️ 🇻🇳 ❤️ 🇻🇳 ❤️

lenajoyeux-star commented on lib/conflicts/conflicts/agents/doctor_agent.py:
```diff
@@ -25,74 +37,51 @@ class DoctorAgent(BaseAgent):
+        if conflict_type not in SPECIALIZED_CONFLICT_TYPES:
+            raise ValueError(
+                f"Invalid conflict_type: {conflict_type}. "
+                f"Must be one of: {SPECIALIZED_CONFLICT_TYPES}"
+            )
+
+        # Load appropriate prompt based on conflict type
+        if conflict_type not in CONFLICT_TYPE_CONFIG:
+            raise ValueError(f"Unknown conflict type: {conflict_type}")
+        prompt_path, agent_name = CONFLICT_TYPE_CONFIG[conflict_type]
```
I don't really get the difference between those two checks

lenajoyeux-star commented on lib/conflicts/conflicts/run_top2.py:
```diff
@@ -0,0 +1,118 @@
+        # Step 3: Save
+        pipeline._save_attempts_to_database(
+            pair_id, document_pair, final_result, attempts, attempt_stats
+        )
```
Is it ok to expose those private functions ?
Should we have a public one that can do the entire process ?

honghanhh replied:
This is a quick script for the subset of 8 examples so I stealed them fast haha. Updated the logic 👍🏻

lenajoyeux-star commented on lib/conflicts/prompts/editor_agent_system.txt:
```diff
@@ -43,34 +49,48 @@ CONFLICT CREATION GUIDELINES:
-    "doc2": {{
-        "op": "delete" | "insert_after" | "replace",
-        "target_text": "exact text from document 2 to modify",
-        "replacement_text": "new text to insert or replace with"
-    }},
-    "conflict_type": "type of medical conflict being created"
+  "doc1": {{
+    "op": "delete" | "insert_after" | "replace",
+    "target_text": "exact text from document 1 to modify",
+    "replacement_text": "new text to insert or replace with"
+  }},
```
Don't we have a contradiction here ? Like, we say "no fence" but then we show the desired output format with fences ?
Or maybe it's not a problem in this case ? I'm not an expert in prompt engineering

honghanhh replied:
You're right haha. However, the parser in base.py already handles both raw JSON and fenced JSON, so this was never a bug. But I remove the fence for the consistency. Thanks.

lenajoyeux-star commented on lib/conflicts/prompts/doctor_agent_biomarker_monitoring_system.txt:
```diff
@@ -0,0 +1,127 @@
+  "reasoning": "explanation of why these proposition pairs were chosen for the {{conflict_type}} conflict, including biomarker analysis, temporal considerations, and why this conflict is sophisticated/clinically nuanced",
+  "modification_instructions": "specific instructions for the Editor Agent on how to create the conflict using the chosen proposition pairs, ensuring biomarker logic and temporal relationships are respected and complexity is prioritized",
+  "editor_instructions": [
+    "1. ANALYZE: Timeframe is [X], {{conflict_type}} conflict is appropriate because [reason related to biomarker trajectory/procedural expectation]",
```
Is this line of the prompt supposed to replace the removed TemporalAnalyser ?
Are we sure this will be efficient ? I don't see a part where we give temporal placeholders to the llm

Plus, DocumentPair still has doc1_timestamp / doc2_timestamp fields, but there are unused now

lenajoyeux-star reviewed (commented):
Great work ! Just some interrogations on my side, once it's cleared then it's a GO for me 🚀

honghanhh committed:
refactor: add conflict_types filter to process_document_pair, pass timestamps to doctor prompts

  - Add optional conflict_types parameter to process_document_pair()
  - Simplify run_top2.py to use the public API instead of internal pipeline methods
  - Pass doc1/doc2 timestamps to all doctor agent prompts for explicit temporal context
  - Remove redundant validation check in DoctorAgent.__init__
  - Remove markdown fences from editor prompt output format example

lenajoyeux-star reviewed (approved):
All good to me !!
